### ISSUE TITLE
FP16 slower than FP32

### ISSUE THREAD

**Issue by KillPinguin** [tensorflow/tensorflow#15585]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Partly
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
RHEL 7
- **TensorFlow installed from (source or binary)**:
unknown
- **TensorFlow version (use command below)**:
('unknown', '1.4.0')
- **Python version**: 
python2.7
- **Bazel version (if compiling from source)**:
unknown
- **GCC/Compiler version (if compiling from source)**:
gcc 4.8
- **CUDA/cuDNN version**:
9.0 / 7.0
- **GPU model and memory**:
Pascal P-100
- **Exact command to reproduce**:
N/A

### Describe the problem
I tried running the tutorial/rnn/ptb and tutorial/images/mnist examples with the --use_fp16 option, to see if speedup was achievable by utilizing the new half-precision features.
It turned out that a single training step for MNIST with FP32 took 3.3ms, with FP16 it was 4ms. For PTB small (I had to use lstm_cell=basic, because other types are not yet supported in FP16), the WPS dropped from 24000 to 22000 when switching to FP16.

So the performance **decreases** when using FP16 in real-world examples.

To check this, I've created a small benchmark for myself (since I can't get the nightly tf build on my machine, I can't run the official benchmarks), which is basically one big matrix multiplication in Tensorflow.

I've run it with square matrices with a width of 8k, 16k and 32k. In each case, FP16 and FP32 yielded nearly the same runtime. Profiling with nvprof I found out that the used CUDA function is **maxwell_fp16_segmemm_fp16_128x128_nn** for FP16 and **sgemm_128x128x8_NN_vec** for FP32.

Since I wanted to double check if matrix multiplication in FP16 is really slower than in FP32 on my GPU, I tried to directly benchmark the GPU using cuBlas with a similar operation. It turns out that here, FP16 is nearly twice as fast as FP32. CuBlas internally uses **maxwell_hgemm_256x128_nn** for matrix multiplication of 16k x 16k square matrices in FP16. (again  according to the nvprof profiler)

So I'm wondering why Tensorflow is unable to achieve similar results in terms of speed, or if I'm doing something wrong in my tests.

### Source code

Tensorflow Code snippet:
```
    graph = tf.Graph()
        with graph.as_default():
          tf_input1 = tf.Variable(tf.truncated_normal([FLAGS.size, FLAGS.size], dtype=get_dtype()))
          tf_input2 = tf.Variable(tf.truncated_normal([FLAGS.size, FLAGS.size], dtype=get_dtype()))
          tf_output = tf.matmul(tf_input1, tf_input2)
  
      with tf.Session(graph=graph) as session:
          tf.global_variables_initializer().run()
          print("Initialized")
          for i in range(FLAGS.times):
              out = session.run([tf_output])#, feed_dict=feed_dict)
          print("Done")
```

cuBlas Code FP16 (Snippet):
```
        uint16_t *d_a;          // d_a - a on the device
        uint16_t *d_b;          // d_b - b on the device
        uint16_t *d_c;          // d_c - c on the device
        cudaStat = cudaMalloc ((void **) &d_a, m * k * sizeof (*a));    // device memory alloc for a
        cudaStat = cudaMalloc ((void **) &d_b, k * n * sizeof (*b));    // device memory alloc for b
        cudaStat = cudaMalloc ((void **) &d_c, m * n * sizeof (*c));    // device memory alloc for c
        stat = cublasCreate (&handle);  // initialize CUBLAS context
        // copy matrices from the host to the device
        stat = cublasSetMatrix (m, k, sizeof (*a), a, m, d_a, m);   //a -> d_a
        stat = cublasSetMatrix (k, n, sizeof (*b), b, k, d_b, k);   //b -> d_b
        stat = cublasSetMatrix (m, n, sizeof (*c), c, m, d_c, m);   //c -> d_c
        uint16_t al = FP_16_ONE;       // al = 1 
        uint16_t bet = FP_16_ONE;      // bet =1
        // matrix - matrix multiplication : d_c = al*d_a *d_b + bet *d_c
        // d_a -mxk matrix , d_b -kxn matrix , d_c -mxn matrix ;
        // al ,bet -scalars

        stat = cublasGemmEx(handle, CUBLAS_OP_N, CUBLAS_OP_N, m, n, k, &al, d_a, CUDA_R_16F, m, d_b, CUDA_R_16F, k, &bet, d_c, CUDA_R_16F, m, CUDA_R_16F, CUBLAS_GEMM_DEFAULT); 

        stat = cublasGetMatrix (m, n, sizeof (*c), d_c, m, c, m);   // cp d_c - >c
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Exact command to reproduce

**Comment by KillPinguin** [comment]
**Exact command to reproduce**
N/A

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by KillPinguin** [comment]
Yes it's still an issue

**Comment by tensorflowbutler** [comment]
The original poster has replied to this issue after the stat:awaiting response label was applied.

**Comment by reedwm** [comment]
In general, fp16 on Pascal GPUs (like your P100) will not be much faster, if faster at all. In your cuBlas example, you pass CUDA_R_16F as the second-to-last parameter, `computeType`, to `cublasGemmEx()`. In TensorFlow, we use fp32 as a compute type, since models do not work well in practice if a lower precision is used as the compute type. This can be changed by setting the evironmental variables [TF_FP16_CONV_USE_FP32_COMPUTE](https://github.com/tensorflow/tensorflow/blob/bffa3e10bf4886f03a68f7e93ba39c91d447f101/tensorflow/stream_executor/cuda/cuda_dnn.cc#L2236) and [TF_FP16_MATMUL_USE_FP32_COMPUTE](https://github.com/tensorflow/tensorflow/blob/de0904e90669d345e0feaa89d1e99b7618fb6677/tensorflow/core/util/matmul_autotune.cc#L44) to 0. But if you do that, the models will probably not train as well.

Note everything I said applies to Pascal. On Volta, fp16 is significantly faster.

Also, I'm not sure where this tutorial/images/mnist example is. Can you clarify what the example is so I can reproduce? Additionally, I cannot run the TensorFlow example you provided, since FLAGs and get_dtype are undefined. Can you post a complete example?

**Comment by KillPinguin** [comment]
@reedwm Thank you for your response.

I did try to set the environmental variables, but it didn't change anything. The called CUDA functions remained the same, and it had no effect on timing.

The tutorials/image/mnist example can be found here: https://github.com/tensorflow/models/tree/master/tutorials/image/mnist
I just used it for trying a prebuilt model with fp16 vs fp32. (again: running with or without the environment variables set didn't change anything)

My code:

```
#!/usr/bin/env python

from __future__ import print_function

import tensorflow as tf
import numpy as np
import argparse

FLAGS = None

def get_dtype():
  return tf.float16 if FLAGS.use_fp16 else tf.float32

def benchmark():
  graph = tf.Graph()
  with graph.as_default():
    tf_input1 = tf.Variable(tf.truncated_normal([FLAGS.size, FLAGS.size], dtype=get_dtype()))
    tf_input2 = tf.Variable(tf.truncated_normal([FLAGS.size, FLAGS.size], dtype=get_dtype()))
    tf_output = tf.matmul(tf_input1, tf_input2)

  with tf.Session(graph=graph) as session:
    tf.global_variables_initializer().run()
    print("Initialized")
    for i in range(FLAGS.times):
      out = session.run([tf_output])
    print("Done")

def parse():
  global FLAGS
  parser = argparse.ArgumentParser()
  parser.add_argument('--use_fp16', default=False, help='use FP16 for benchmarking', action='store_true')
  parser.add_argument('--size', default=4096, type=int, help='size of matrices to multiply')
  parser.add_argument('--times', default=10, type=int, help='amount of multiplications')
  FLAGS, unparsed = parser.parse_known_args()

if __name__ == "__main__":
  parse()
  benchmark()
```

What I did:
```
export TF_FP16_CONV_USE_FP32_COMPUTE=0
export TF_FP16_MATMUL_USE_FP32_COMPUTE=0
nvprof ./Matrixmul.py --use_fp16 --size 16000
```
Output:
```
[...]
==56407== Profiling application: python ./Matrixmul.py --use_fp16 --size 16000
==56407== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   96.94%  8.16365s        10  816.37ms  814.54ms  819.59ms  maxwell_fp16_sgemm_fp16_128x128_nn
```


Without the environment variables:
```
export TF_FP16_CONV_USE_FP32_COMPUTE=1
export TF_FP16_MATMUL_USE_FP32_COMPUTE=1
nvprof ./Matrixmul.py --use_fp16 --size 16000
```
Output:
```
==70743== Profiling application: python ./Matrixmul.py --use_fp16 --size 16000
==70743== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   96.49%  8.17353s        10  817.35ms  814.50ms  824.69ms  maxwell_fp16_sgemm_fp16_128x128_nn
```


With FP32:
```
nvprof ./Matrixmul.py --size 16000
```
Output:
```
[...]
==89305== Profiling application: python ./Matrixmul.py --size 16000
==89305== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   93.54%  8.14831s        10  814.83ms  810.68ms  818.11ms  sgemm_128x128x8_NN_vec
```

Note that **maxwell_fp16_sgmemm_fp16_128x128_nn** uses only FP16 storage but computes in FP32. So why is this function called when explicitly setting the TF_FP16_MATMUL_USE_FP32_COMPUTE to 0?

**Comment by reedwm** [comment]
Looking at [matmul_op.cc](https://github.com/tensorflow/tensorflow/blob/de0904e90669d345e0feaa89d1e99b7618fb6677/tensorflow/core/kernels/matmul_op.cc#L388), it looks like we sometimes do not pass the compute type from `TF_FP16_MATMUL_USE_FP32_COMPUTE` to cuBLAS. Adding a log statement, this seems to be occurring. /CC @yzhwang can you comment?

Looking at tutorials/image/mnist, I'm not sure why fp16 is slower. When I generate Timelines for that example, the overhead of generating a timeline causes both to take the same amount of time, so it's hard to debug this. Since the example is so simple and fp16 has little performance benefit from P100s anywhere, it's not worth spending more time to debug this.

**Comment by yzhwang** [comment]
Note that to call cublasGemmEx, one needs to enable autotune for matmul, which can be done by setting this environment var:
`export TF_MATMUL_AUTOTUNE_ENABLE = 1`
By default the autotune for matmul is disabled, 

So first you can try turning on the autotune for matmul.

@reedwm For your question, I think the computation_type is the param that is sent in. I do not see any possibility that it failed to get sent to cublas. I'd set it with:
`os.environ['TF_FP16_MATMUL_USE_FP32_COMPUTE']=1`
in python just to make sure it gets set.

**Comment by KillPinguin** [comment]
@yzhwang 
setting TF_MATMUL_AUTOTUNE_ENABLE also does not change the behavior.
Code:
```
$ export TF_MATMUL_AUTOTUNE_ENABLE=1
$ export TF_FP16_MATMUL_USE_FP32_COMPUTE=0
$ python -c "import os; print os.environ['TF_FP16_MATMUL_USE_FP32_COMPUTE']"
0
$ python -c "import os; print os.environ['TF_MATMUL_AUTOTUNE_ENABLE']"
1
$ nvprof ./Matrixmul.py --size 16000 --use_fp16
```
Output:
```
[...]
==126141== Profiling application: python ./Matrixmul.py --size 16000 --use_fp16
==126141== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   96.77%  8.17152s        10  817.15ms  816.77ms  817.74ms  maxwell_fp16_sgemm_fp16_128x128_nn
[...]
```

**Comment by KillPinguin** [comment]
@yzhwang 

According to matmul_op.cc it doesn't seem to matter anyways, because autotune is not supported for halfs:
https://github.com/tensorflow/tensorflow/blob/b06326cadc7b45078291e8fb3da0b7e941978540/tensorflow/core/kernels/matmul_op.cc#L235

That's why you don't autotune here https://github.com/tensorflow/tensorflow/blob/b06326cadc7b45078291e8fb3da0b7e941978540/tensorflow/core/kernels/matmul_op.cc#L299

And launch the normal Gemm (without Ex) here: https://github.com/tensorflow/tensorflow/blob/b06326cadc7b45078291e8fb3da0b7e941978540/tensorflow/core/kernels/matmul_op.cc#L398

**Comment by yzhwang** [comment]
@KillPinguin Thanks for pointing out. This was a change I made a couple of months ago due to a cublas bug back then. I will try to fix the issue in two steps:
1) enable autotune support for fp16. I need to do some tests to make sure that no bug shows up.
2) enable autotune for matmul by default in general. Last time I tried to enable it by default I got mixed performance results (for some shapes autotune helps, for some shapes it hurts the performance).

Both step involves simple change but extensive tests/benchmarks, so do not expect the change to happen within weeks.

Also, just so you know, as long as the first step is done, you should see a difference:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/cuda/cuda_blas.cc#L1863
since matmul for half is calling SgemmEx() now, just the compute type is still FP32.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by cpandar** [comment]
Hi @yzhwang & @KillPinguin - we've been noticing the same issues, and were wondering if you had any new thoughts or advice on this? Thanks in advance.

**Comment by bhack** [comment]
Any news?

**Comment by KillPinguin** [comment]
@cpandar we now also tried it on a V100 lately, which also was slower with FP16 than FP32.

Unfortunately I have switched positions since then and am no longer involved into the project, so I can't  conduct any more tests.

**Comment by jart** [comment]
This appears to be a duplicate of #5592. Please follow that issue. If you feel this issue should be reopened, please provide new evidence and I'll do that.

**Comment by alvaroslm** [comment]
> This appears to be a duplicate of #5592. Please follow that issue. If you feel this issue should be reopened, please provide new evidence and I'll do that.

That issue is closed but this problem was never fixed. From simple test programs carrying out a few ops on a graph to several tf.keras layers, anything with float16 I've tried is significantly slower than 32-bit fp on a Turing GPU.

**Comment by tfboyd** [comment]
With the new FP16 we are seeing speedups with Keras us TF 2.0. It is still
experimental.  With a single gpu and synthetic data we are pushing 1,300
images per second.  More to come and working on the input pipeline (real
data).

**Comment by boiihong** [comment]
@tfboyd 
Hello, Would you give more specific information about the experiment?
I am comparing performance of FP32 and FP16 on the simple 2-layer LSTM model.
I have built the latest version of tensorflow + keras and implemented the application on that.

i couldn't see any acceleration with FP16, yet.
And i made a issue thread regarding this at #26500

**Comment by alvaroslm** [comment]
> @tfboyd
> Hello, Would you give more specific information about the experiment?
> I am comparing performance of FP32 and FP16 on the simple 2-layer LSTM model.
> I have built the latest version of tensorflow + keras and implemented the application on that.
> 
> i couldn't see any acceleration with FP16, yet.
> And i made a issue thread regarding this at #26500

Same here, lstm inference at least doesn't seem to improve. I did manage to run the tf cnn benchmark (after patching it, because it tries to include some stuff that's no longer there under the same name) and it does perform as expected. But it seems to me that fp16 support is spotty outside cnns

**Comment by tfboyd** [comment]
I failed to read the entire thread, sorry about that.  I do not have any LSTM examples FP16 yet.  We are just finishing up easy FP16 with NVIDIA to be released in a few weeks and per usual the first focus is on convs.

**Comment by reedwm** [comment]
@protoget do you know if float16 should be significantly faster on LSTM inference?

**Comment by IvyGongoogle** [comment]
meet the same problem on v100: fp16 is very slow than fp32

**Comment by UndeadBlow** [comment]
I faces the same issue. 
I converted model to pure DT_HALF, run Google Cloud with Nvidia Tesla T4 and model with DT_HALF is slower in about 8-9 times than DT_FLOAT.

Is this normal behaviour? Did anyone succeed to run FP16 with Tensorflow at least somewhere?

Maybe FP16 can be used with Tensorflow TensorRT binding?

**Comment by LunaonMSK** [comment]
meet the same problem with v100 on aws p3 instance,fp16 is 3x slower than fp32,I have not found the solution

**Comment by congchan** [comment]
how to check if I am using fp32 or fp16 on v100?
